### PR TITLE
Compare table names using prefix and suffix

### PR DESCRIPTION
**Pull request by VladyslavKochetkov**
### Summary:

I ran into an issue when using models and transactions where if you used a suffix or prefix (in my case suffix dev or prod) and tried to use transactions it would fail with the error message...

```
Table "table_name_{SUFFIX}" not found. Please register the table with dynamoose before using it in transactions."
```

#### Model
```ts
const AnonContactModel = dynamoose.model<AnonContact>(
  'anon_contact_test1',
  {
    PK: {
      default: `ac:${uuid()}`
      type: String,
      hashKey: true,
    },
    SK: {
      type: String,
      rangeKey: true,
      index: {
        type: 'global',
        name: 'SK-PK',
        rangeKey: 'PK',
        throughput: 'ON_DEMAND',
    },
    type: {
      type: String,
      enum: ['email', 'phone'],
    },
  },
  {
    suffix: process.env.STAGE, // "dev" or "prod"
    throughput: 'ON_DEMAND',
  }
)
```

#### General
```ts
const count = await AnonContactModel.query()
    .where('SK')
    .eq(`ac:${hashContact(input.email)}`)
    .using('SK-PK')
    .count()
    .exec();
console.log(count) // 0 or 1 - this works perfectly fine

await dynamoose.transaction([
  AnonContactModel.transaction.create({ SK: input.email, type: 'email' }),
  AnonContactModel.transaction.create({ SK: input.phoneNumber, type: 'phone' }),
  // any other transaction with a model created using a suffix
]

/* 

  !! ERROR !!

  Table "anon_contact_dev" not found. Please register the table with dynamoose before using it in transactions.

*/
```

### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

- [ ] Test added to report bug (GitHub issue #---- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [ ] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [ ] I have updated the Dynamoose documentation (if required) given the changes I made
- [ ] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/main/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

**Comment by famence** [comment]
+1. Also facing this issue. Needs to be merged ASAP.
This is not a breaking change, but a hot fix

**Comment by famence** [comment]
@VladyslavKochetkov I would like to make some improvements, but i dont know how to propose changes to your pr, so i have made a pr to your pr 😄


Please check it out — https://github.com/VladyslavKochetkov/dynamoose/pull/1

All tests passed.

**Comment by VladyslavKochetkov** [comment]
Thanks @famence - merged your changed into my branch.

File: packages/dynamoose/lib/Model/index.ts [hunk 1]
```diff
@@ -14,6 +14,7 @@ import {AttributeMap} from "../Types";
 import * as DynamoDB from "@aws-sdk/client-dynamodb";
 import {GetTransactionInput, CreateTransactionInput, DeleteTransactionInput, UpdateTransactionInput, ConditionTransactionInput} from "../Transaction";
 import {Table, TableOptionsOptional} from "../Table";
+import {custom as customDefaults, original as originalDefaults} from "../Table/defaults";
```
Comment (fishcharlie) [inline packages/dynamoose/lib/Model/index.ts:17]: Since these are table specific lets include table in the variable names.

File: packages/dynamoose/lib/Model/index.ts [hunk 2]
```diff
@@ -104,7 +105,7 @@ export interface ModelTableOptions extends TableOptionsOptional {
 
 interface ModelInternalProperties {
 	name: string;
-	options: TableOptionsOptional;
+	options: ModelTableOptions;
```
Comment (fishcharlie) [inline packages/dynamoose/lib/Model/index.ts:108]: What is the purpose of this change?

File: packages/dynamoose/lib/Model/index.ts [hunk 3]
```diff
@@ -407,6 +406,12 @@ export class Model<T extends ItemCarrier = AnyItem> extends InternalPropertiesCl
 		return this.getInternalProperties(internalProperties).name;
 	}
 
+	get tableName (): string {
```
Comment (fishcharlie) [inline packages/dynamoose/lib/Model/index.ts:409]: Slightly confused about the purpose behind this.

File: packages/dynamoose/lib/Model/index.ts [hunk 4]
```diff
@@ -114,7 +115,6 @@ interface ModelInternalProperties {
 	getHashKey: () => string;
 	getRangeKey: () => string | void;
 	table: () => Table;
-	tableName: string;
```
Comment (fishcharlie) [inline packages/dynamoose/lib/Model/index.ts:117]: Confused about the purpose behind this change.

File: packages/dynamoose/lib/utils/dynamoose/returnModel.ts [hunk 5]
```diff
@@ -9,7 +9,7 @@ export default <T extends Item = AnyItem>(model: Model<T>): ModelType<T> => {
 		Object.keys(model),
 		Object.keys(Object.getPrototypeOf(model)),
 		Object.getOwnPropertyNames(Object.getPrototypeOf(model))
-	]).filter((key) => !["constructor", "name"].includes(key));
+	]).filter((key) => !["constructor", "name", "tableName"].includes(key));
```
Comment (fishcharlie) [inline packages/dynamoose/lib/utils/dynamoose/returnModel.ts:12]: What is the purpose of this change?

**Review by fishcharlie — CHANGES_REQUESTED** [review]
What about unit tests as well?

File: packages/dynamoose/lib/Model/index.ts (hunk 2 above)
Comment (famence) [reply packages/dynamoose/lib/Model/index.ts:108]: interface ModelTableOptions extends TableOptionsOptional and is [a type of Model.contructor thrird argument "options"](https://github.com/dynamoose/dynamoose/pull/1594/files/88309e897166d0df225b4f4e9ef21bbc7e164f41#diff-1d6224e89f76a90eae897c943da376eda750143ddf3e19912f4beaba46768893R194)

Since ModelInternalProperties.options used to store this "options" value, it will be correct if it has the same type.

File: packages/dynamoose/lib/Model/index.ts (hunk 4 above)
Comment (famence) [reply packages/dynamoose/lib/Model/index.ts:117]: Just moved tableName property of model from ModelInternalProperties to Model properties

File: packages/dynamoose/lib/Model/index.ts (hunk 2 above)
Comment (fishcharlie) [reply packages/dynamoose/lib/Model/index.ts:108]: @famence But why is it necessary? Why does `TableOptionsOptional` not work in this case?

File: packages/dynamoose/lib/Model/index.ts (hunk 4 above)
Comment (fishcharlie) [reply packages/dynamoose/lib/Model/index.ts:117]: @famence Ok, but why?

File: packages/dynamoose/lib/utils/dynamoose/returnModel.ts (hunk 5 above)
Comment (famence) [reply packages/dynamoose/lib/utils/dynamoose/returnModel.ts:12]: Due to moving tableName property from ModelInternalProperties to root of Model class, it must be filtered here just like the "name" property

File: packages/dynamoose/lib/Model/index.ts (hunk 3 above)
Comment (famence) [reply packages/dynamoose/lib/Model/index.ts:409]: Model.tableName is now a getter instead of an optional string parameter of ModelInternalProperties.
The code of this getter allows to always get the correct name of the model table, even if the table itself has not been initialised yet

**Comment by VladyslavKochetkov** [comment]
@fishcharlie Would you be more comfortable merging the first commit, if I wrote unit tests for the added function?

**Comment by fishcharlie** [comment]
@VladyslavKochetkov If this PR only includes fdd84fdd085b37e07d90212b05dfbc6d156ba985 + automated tests, yes I'd merge that probably if all the CI checks pass. I don't see any issues with fdd84fdd085b37e07d90212b05dfbc6d156ba985.
